333 add buttons for task tracking apps

tansdf opened the pull request:
Implemented Jira integration via Basic Auth
- Introduced Basic Authentication for Jira Cloud, allowing users to create tasks from incidents.
- Updated environment configuration to support Jira API token and user email for authentication.
- Introduced a new JinjaTemplate class to facilitate message and notification rendering within the IMPulse application.
- Implemented methods for forming messages and notifications using Jinja2 templates.
- Updated application modules to utilize the new JinjaTemplate class, enhancing template management and rendering capabilities.
- Added a new function `load_template_file` to load Jinja templates from the filesystem, improving template management.
- Updated the `JiraIntegration` class to utilize the new template loading function for Jira issue summary and description templates.
- Introduced two new Jinja templates: `jira_summary.j2` and `jira_description.j2` for better incident reporting in Jira.
- Added unit tests for the new template loading functionality and the JinjaTemplate class to ensure reliability and correctness.

tansdf committed:
Implement Jira integration for incident task management

- Added support for Jira integration, allowing users to create and manage tasks directly from incidents.
- Introduced OAuth authentication flow for Jira, including endpoints for authorization and callback handling.
- Enhanced the application configuration to include Jira-specific settings such as project key and OAuth credentials.
- Updated application classes (Mattermost, Slack, Telegram) to include Jira buttons for task creation and management.
- Implemented Jira client classes for both Cloud and Server versions, handling authentication and task creation logic.
- Added tests to ensure proper functionality of the Jira integration and its interaction with incidents.

tansdf committed:
[TMP] Add Jira integration documentation and implementation status

- Introduced comprehensive documentation for Jira integration, covering both Cloud (OAuth 2.0) and Server (OAuth 1.0a) setups.
- Added example environment variable configurations for both Jira types.
- Documented the implementation status, detailing completed features for Jira Cloud and outlining required implementations for Jira Server.
- Included troubleshooting tips, security considerations, and a testing checklist to assist users in setting up and verifying the integration.
- Enhanced overall documentation structure for clarity and ease of use.

tansdf committed:
Refactor Jira integration to use Basic Authentication

- Removed OAuth 2.0 and OAuth 1.0a implementations for Jira integration.
- Introduced Basic Authentication for Jira Cloud, allowing users to create tasks from incidents.
- Updated environment configuration to support Jira API token and user email for authentication.
- Refactored Jira client to handle task creation using Basic Auth.
- Removed related OAuth endpoints and documentation, streamlining the integration process.
- Enhanced application classes to support the new authentication method and ensure compatibility with existing functionality.

tansdf committed:
[TMP] Reference docs

tansdf committed:
Add Jira integration with Basic Authentication and template rendering

- Introduced a new JinjaTemplate class for rendering messages and notifications.
- Implemented JiraClient for interacting with the Jira Cloud API using Basic Authentication.
- Created JiraIntegration class to manage task creation from incidents, utilizing Jinja templates for issue summaries and descriptions.
- Updated application classes to support the new Jira integration and removed the deprecated Jira module.
- Enhanced tests for JiraClient and JiraIntegration to ensure proper functionality and error handling.
- Updated environment configuration to include Jira credentials for testing purposes.

tansdf committed:
Add JinjaTemplate class for rendering messages and notifications

- Introduced a new JinjaTemplate class to facilitate message and notification rendering within the IMPulse application.
- Implemented methods for forming messages and notifications using Jinja2 templates.
- Added Jira-specific templates for summarizing and describing incidents.
- Updated application modules to utilize the new JinjaTemplate class, enhancing template management and rendering capabilities.

tansdf committed:
Enhance Jinja template functionality and integrate Jira templates

- Added a new function `load_template_file` to load Jinja templates from the filesystem, improving template management.
- Updated the `JiraIntegration` class to utilize the new template loading function for Jira issue summary and description templates.
- Introduced two new Jinja templates: `jira_summary.j2` and `jira_description.j2` for better incident reporting in Jira.
- Added unit tests for the new template loading functionality and the JinjaTemplate class to ensure reliability and correctness.

DiTsi commented on app/im/mattermost/config.py:
```diff
@@ -25,6 +25,16 @@
             'style': 'default'
         }
+    },
+    'jira': {
```
Instead of "jira" we should use "task" word, because we will support another task management tools (just one for every installation).

tansdf replied:
The task word is too wide. At this point it doesn't make any sense to chage it. But i'd suggest at least `"file_ticket"`

DiTsi commented on app/im/mattermost/threads.py:
```diff
@@ -17,8 +17,66 @@ def chain_attrs(chain_enabled, status):
+        if task_link:
+            # If task exists, button opens the link
+            actions.append({
+                "id": "jira",
```
"task". Same below

DiTsi commented on app/im/slack/slack_application.py:
```diff
@@ -136,15 +136,19 @@ async def buttons_handler(self, payload, incidents, queue_, route):
                     logger.info(f'Incident {incident_.uuid} -> button STATUS pressed (enabled)')
                     incident_.status_enabled = True
+            elif action['name'] == 'jira':
+                logger.info(f'Incident {incident_.uuid} -> button JIRA pressed')
```
```suggestion
                logger.info(f'Incident {incident_.uuid} -> button Task pressed')
```

tansdf committed:
Refactor Mattermost and Slack integration to streamline action button creation

- Introduced `build_mattermost_actions` and `build_slack_actions` functions to centralize the logic for creating action buttons in Mattermost and Slack messages.
- Removed redundant code and improved readability by consolidating button creation logic.
- Updated relevant functions to utilize the new action building methods, ensuring consistent button configurations across different message types.
- Enhanced integration with Jira by allowing optional task links in action buttons.

tansdf committed:
Refactor action handling in application classes to improve code organization

- Introduced helper methods `_handle_chain_action`, `_handle_status_action`, and `_handle_jira_action` in Mattermost, Slack, Telegram, and Application classes to streamline action button processing.
- Enhanced the `buttons_handler` method in each application class to utilize these new methods, reducing code duplication and improving readability.
- Added a mechanism to track async tasks in the Application class to prevent premature garbage collection.
- Updated unit tests to reflect changes in action handling logic.

tansdf committed:
Refactor Jira integration to Task Management integration

- Renamed references from Jira integration to Task Management integration across the application, including configuration checks and integration handling.
- Updated action handling methods in application classes to reflect the new naming convention, improving clarity and consistency.
- Adjusted button configurations in Mattermost, Slack, and Telegram to replace Jira buttons with File Ticket buttons, aligning with the new integration focus.
- Modified unit tests to accommodate changes in integration naming and functionality.

tansdf committed:
Remove Jira integration documentation files

- Deleted `JIRA_INTEGRATION.md` and `JIRA_QUICK_START.md` as part of the transition to a new Task Management integration.
- This cleanup aligns with the recent refactor to streamline the integration process and improve overall documentation clarity.

tansdf committed:
Enhance NullApplication update method to include optional task_link parameter

- Updated the `update` method in the `NullApplication` class to accept an additional `task_link` parameter, improving flexibility for future integrations.
- This change prepares the class for potential enhancements related to task management without altering existing functionality.

DiTsi committed:
remove dot

DiTsi committed:
update task button logs and slack docs

tansdf committed:
Refactor ticket button handling across applications

- Renamed `file_ticket` references to `ticket` in Mattermost, Slack, and Telegram configurations and action handling methods for consistency.
- Updated button handling methods to reflect the new naming convention, improving clarity in the codebase.
- Adjusted related unit tests to ensure proper functionality with the new button naming.

tansdf committed:
Refactor ticket button handling in Mattermost, Slack, and Telegram

- Removed the 'open' button configuration for the ticket button across Mattermost, Slack, and Telegram to streamline button functionality.
- Updated action handling methods to ensure the ticket button only appears when a task link is not provided, enhancing user experience.
- Adjusted related unit tests to verify the correct behavior of the ticket button under the new configuration.

tansdf committed:
Implement status and ticket action handling in Application class

- Added `_handle_status_action` and `_handle_ticket_action` methods to the Application class to manage status-related and ticket button actions across different integrations.
- Updated Mattermost, Slack, and Telegram application classes to utilize the new action handling methods, improving code reuse and consistency.
- Enhanced logging for button actions to provide clearer insights into user interactions.
- Introduced a utility function in tests to standardize button configurations for Telegram, reducing code duplication in test cases.

tansdf committed:
Implement message update handling in Jira integration and queue manager

- Changed the queue item type from 'update_status' to 'update_message' in the Jira integration to allow for message updates without altering incident status.
- Added a new `MessageUpdateHandler` to manage message updates in the queue manager, ensuring that incident status remains unchanged during updates.
- Updated unit tests to verify the correct handling of message updates and the new functionality in the Jira integration.

tansdf committed:
Refactor action handling to remove async from status and ticket methods

- Changed `_handle_status_action` and `_handle_ticket_action` methods in the Application class from async to synchronous, improving performance and simplifying the call structure.
- Updated Mattermost, Slack, and Telegram application classes to reflect these changes, ensuring consistent action handling across integrations.

tansdf committed:
Add task_creation_in_progress flag to Incident and update Jira integration handling

- Introduced a new boolean flag `task_creation_in_progress` in the Incident class to manage the state of task creation.
- Updated the JiraIntegration class to check this flag before creating a task, preventing duplicate task creation attempts.
- Adjusted the handling of task creation responses to ensure the flag is reset appropriately after task creation or failure.
- Modified unit tests to verify the correct behavior of the new flag and its integration with the Jira task creation process.

DiTsi committed:
Enhance task management integration and refactor related components

- Added support for task management configuration in the application, allowing integration with Jira for task creation.
- Updated the `get_application` function to accept a `task_management_config` parameter, enabling dynamic task management setup.
- Refactored button handling across Mattermost, Slack, and Telegram to replace 'ticket' references with 'task', improving consistency in terminology.
- Introduced new template handling for task creation in the Jira integration, allowing custom summary and description templates.
- Updated documentation to include task management configuration details and environment variables.
- Adjusted unit tests to reflect changes in task handling and ensure proper functionality across integrations.

DiTsi committed:
Switch to Jira API v2, update description and summary templates

DiTsi committed:
Refactor task management checks to include application configuration

- Updated task management conditionals in Mattermost, Slack, and Telegram integrations to utilize the new application configuration for task management.
- Enhanced documentation to clarify the optional Task button functionality and its integration with task management systems.

DiTsi committed:
fix jira templates probles and update tests

DiTsi committed:
split huge functions

DiTsi committed:
update test
